transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java

```diff
@@ -671,13 +674,41 @@ public void run() {
 
     private static void safeSuccess(ChannelPromise promise) {
         if (!(promise instanceof VoidChannelPromise) && !promise.trySuccess()) {
-            logger.warn("Failed to mark a promise as success because it is done already: {}", promise);
+            Throwable err = promise.cause();
+            if (err == null) {
+                logger.warn("Failed to mark a promise as success because it is done already: {}", promise);
+            } else {
+                logger.warn("Failed to mark a promise as success because it is done already: {} => {}",
+                        promise, stackTraceToString(err));
+            }
         }
     }
 
     private static void safeFail(ChannelPromise promise, Throwable cause) {
         if (!(promise instanceof VoidChannelPromise) && !promise.tryFailure(cause)) {
-            logger.warn("Failed to mark a promise as failure because it's done already: {}", promise, cause);
+            Throwable err = promise.cause();
+            if (err == null) {
+                logger.warn("Failed to mark a promise as failure because it's done already: {} ", promise, cause);
+            } else {
+                logger.warn("Failed to mark a promise as failure because it's done already: {} => {} ",
+                        promise, stackTraceToString(err), cause);
+            }
+        }
+    }
+
+    private static String stackTraceToString(Throwable cause) {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        PrintStream pout = new PrintStream(out);
+        cause.printStackTrace(pout);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Use a logger to log this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1148)